### PR TITLE
"Use the whole buffer" is !0, not 0

### LINKED ISSUE #654: Change size = 0 from meaning the whole remaining size
Currently, providing size = 0 means the whole size (or the remaining size). This makes it impossible to support zero-sized bindings, see https://github.com/gpuweb/gpuweb/issues/686
wgpu-rs is now using `Option<>` for size, as of https://github.com/gfx-rs/wgpu-rs/pull/302.
We should change the logic here to have `!0` instead being the special value for the whole size.

### PR DESCRIPTION
**Pull request by paulkernfeld**
A first step towards #654
Applies to BufferBinding, set_vertex_buffer, set_index_buffer

### CODE REVIEW

**Comment by paulkernfeld** [comment]
If this needs to be "backwards compatible," I think it can be done in three parts instead of two:

1. Do this PR, but with `WHOLE_SIZE = 0`
2. Update `wgpu-rs` to use the `WHOLE_SIZE` constant
3. Update `wgpu` with `WHOLE_SIZE = !0`

File: wgpu-core/src/command/render.rs [hunk 1]
```diff
@@ -1065,15 +1065,15 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
...
+                    state.vertex.inputs[slot as usize].total_size = if size != crate::WHOLE_SIZE {
                         size
                     } else {
                         buffer.size - offset
                     };
 
                     let range = hal::buffer::SubRange {
                         offset,
-                        size: if size != 0 { Some(size) } else { None },
+                        size: if size != crate::WHOLE_SIZE { Some(size) } else { None },
```
[the remark this answers is not in the document]
Comment (paulkernfeld) [reply wgpu-core/src/command/render.rs]: Oops! I force-pushed a `fmt` fix.

**Comment by kvark** [comment]
Bummer, nightly broke something:
>  error: legacy asm! syntax is no longer supported
  --> /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/generator-0.6.20/src/detail/x86_64_unix.rs:19:9

**Comment by paulkernfeld** [comment]
I made an issue for the CI failure (#664)

**Comment by paulkernfeld** [comment]
> My only worry is whether you covered all the cases with this PR, or not.

Yeah, good question. I'm not sure because "I don't know what I don't know." My techniques were:

1. Look for code used by https://github.com/gfx-rs/wgpu-rs/pull/302
2. Search for the string `.size`
3. Search for strings `!= 0` and `== 0`

Possible places that might need changes:

- `BufferDescriptor`
- `map_buffer`
- `create_buffer`
- `device_create_bind_group`

**Comment by kvark** [comment]
Ok, great! The CI is the only blocker left to land this.
We can either wait for any info from Rust team, or change the CI to allow failures on nightly, temporarily.

**Comment by kvark** [comment]
Just figured out a problem with this PR that would be very good to solve early.
Our API tracing currently has stuff like this:
```
            SetVertexBuffer(
                slot: 0,
                buffer_id: Id(6, 1, Vulkan),
                offset: 0,
                size: 0,
            ),
```
If we merge the PR as is, it will become a bit creepy:
```
            SetVertexBuffer(
                slot: 0,
                buffer_id: Id(6, 1, Vulkan),
                offset: 0,
                size: 18446744073709551615,
            ),
```
So we need to make something smarter than this.

**Comment by kvark** [comment]
I don't have super bright ideas on how to get the serialization right for the buffer sizes, but here are some:
  - introduce a separate type `i64` for the size and treat -1 as the remaining size
  - patch the serializer to treat these separately (hard)
  - leave as is, and suffer

**Comment by paulkernfeld** [comment]
Could we make a newtype that wraps a `u64` or `i64`, maybe?

What about just using an `Option`? The code would probably be easiest to understand and work with that way.

A more ambitious idea would be to make a `BasicallyU64` type that allows users to jam arbitrary info into the higher significant bits. Actually maybe you could just use `bitflags` and a newtype to do this.

**Comment by kvark** [comment]
There are multiple considerations here:
  1. Large surface of wgpu needs to be FFI-friendly. So if the C API uses any structure having that "size", turning it from `u64` into `Option<u64>`  would be highly inconvenient: all FFI clients would have to manually convert and introduce their duplicating structures. This can still be OK, e.g. the bind groups already use enums, and they are converted by native clients. We just have to be very careful, and check how [wgpu-native](https://github.com/gfx-rs/wgpu-native/), for example, builds with the changes.
  2. Size affects performance. For example, all pass commands are serialized. If we start serializing `Option<u64>` that may end up wasting 8 bytes, affecting all runtime.

> allows users to jam arbitrary info into the higher significant bits

Not sure how this would work. Perhaps, you could provide more details?

**Comment by paulkernfeld** [comment]
> Large surface of wgpu needs to be FFI-friendly

Oops, I forgot about FFI 😨

I pushed up some code that might address some of the concerns. It uses a transparent repr so that FFI users can still just pass an integer, but then it uses a serialization "buddy" type so that the serialized format looks cleaner.

> 2. Size affects performance. For example, all pass commands are serialized. If we start serializing `Option<u64>` that may end up wasting 8 bytes, affecting all runtime.

I don't think that I have solved this issue. Is there a different "buddy type" that we could use to solve the problem?

File: wgpu-core/src/lib.rs [hunk 2]
```diff
@@ -54,6 +56,56 @@ type Epoch = u32;
...
+
+impl BufferSize {
+    const WHOLE: BufferSize = BufferSize(!0u64);
+}
+
+/// This type allows us to make the serialized representation of a BufferSize more readable
+#[allow(dead_code)]
+#[cfg_attr(feature = "trace", derive(serde::Serialize))]
+#[cfg_attr(feature = "replay", derive(serde::Deserialize))]
+enum SerBufferSize {
```
Comment (kvark) [inline wgpu-core/src/lib.rs:83]: could we move this enum and its `From` implementations into `device::trace` module, where it's only used?

Comment (paulkernfeld) [reply wgpu-core/src/lib.rs]: 👍

**Pull request merged**